subsys/bluetooth/mesh/settings.c

```diff
@@ -183,7 +130,86 @@ static void store_pending(struct k_work *work)
 	}
 }
 
+static int mesh_commit(void)
+{
+	if (!bt_mesh_subnet_next(NULL)) {
+		/* Nothing to do since we're not yet provisioned */
+		return 0;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
+		bt_mesh_proxy_prov_disable(true);
+	}
+
+	bt_mesh_net_settings_commit();
+	bt_mesh_model_settings_commit();
+
+	atomic_set_bit(bt_mesh.flags, BT_MESH_VALID);
+
+	bt_mesh_start();
+
+	return 0;
+}
+
+static int mesh_export(int (*export_func)(const char *name, const void *val,
+					  size_t val_len))
+{
+	BT_DBG("");
+
+	/* Use settings_save_one because export_func is actually settings_save_one */
```

Comment:
@nvlsianpu @Laczen @jhedberg @trond-snekvik @alxelax 

We discussed this with @nvlsianpu offline, but I would like you guys to get your opinion on this.

Basically, `export_func` is `settings_save_one`, but implementing it this way may create a potential issue in future in case the settings subsystem is rewritten.

The more proper way of doing that would be to pass a pointer of either `export_func` or `settings_save_one` into `store_pending`. However, it will require extending all the underlying functions...

Reply:
@PavelVPV, I don't quite get what you need the opinion on. I can give you some background on the export functionality: the idea was to provide a way to report the settings, this could be to store them to flash, but it could also be to report them to a user interface or ... 

Doing a settings_save_one is only one of the possibilities.

As it is unknown if this export is save or not (it could be a dump to screen) it is up to the designer to allow or disallow the export functionality by either specifying resp. "nulling" the export function. Most mesh functions do not have a export functionality because it is considered unsafe to share the information.

I don't know if this answers your question, but I would consider it unwise for all mesh settings variables to provide a export function.

Reply:
@Laczen, thanks for your reply!

When you are talking about export functionality, do you mean implementation of `h_export` or usage of `export_func`?

I didn't get about the unwise part, because settings subsystem documentation clearly defines what `h_export` is used for and what it uses as `export_func`:
```
A call to settings_save() uses an h_export implementation to store different data in one operation using settings_save_one().
```
Also, do you mean that using `settings_save_one` is not safe either? That sounds like using settings subsystem is unsafe for mesh at all.

What I am trying to emphasise (and asking an opinion for) is that the current design (providing `export_func` in `h_export`) makes some problems in the implementation, which I would like to avoid. On the other hande, I may wrongly understand settings subsystem API (the purpose of `h_export` handler), but then it should be clearly written in the documentation.

Reply:
@PavelVPV, generally, this approach will work but it has a kind of logical inconsistency.
Looks like `h_export` functionality has been designed really to export data to wherever. If it is used for immediate storing everything for the power down case it has to clean up invalidated entries as well. Otherwise, after power-up, they will be restored.  That means it not only stores settings but cleans them either.    
Furthermore, it is unsafe to rely on an external function pointer for storing data, including security credentials. Sounds like a potential security vulnerability.

Reply:
@PavelVPV, as shown in the documentation the `settings_save()` uses an `h_export` implementation to store different data in one operation. This however only works when a `h_export` routine is defined for each `settings_handler`, if there is no `h_export` nothing will be saved.

There used to be a `settings_export()` that you could use to do whatever you wanted with the name and value (e.g. dump to screen). This seems to have been removed along the way.

At the moment the export functionality seems only used for settings_save(). So you use `h_export()` for this, but it might not be easy, each settings_handler is used to retrieve multiple stored items, the settings_handler `h_set` is responsible for decoding the name into the appropiate set routine. For the settings_save() you need to do the opposite and create multiple h_export routines for whatever needs storing.

If your goal is only to ensure that the latest value are written to flash it would be easier to find a way to set the next storage `time` to `now` (and reschedule), or to define a flush to disk routine and call that either directly or based on some other trigger.

Reply:
@Laczen, thanks for the clarification, that surprises me. Seems like the only way to achieve the goal is to do it the way you described in your comment.

Reply:
@PavelVPV, I just went over your code again (should have done this more thoroughly before) and your proposed change might break things. The idea behind `settings_save()` is to allow the user to save all settings. Your change only saves very limited amount of data.

I think it would be acceptable to define `settings_save()` as a routine that would allow to save any unsaved data to flash using the `h_export` routine.

Reply:
@alxelax I'm not sure I got you. `settings_save_one` or `settings_delete` will be called if any of `bt_mesh_settings_flag` flag is set, which in turns will store or delete pending data.

Reply:
@Laczen I think it will save any mesh data for which `bt_mesh_settings_flag` flag is set.

Reply:
> @Laczen I think it will save any mesh data for which `bt_mesh_settings_flag` flag is set.

No need, see my comment above.

Reply:
> @Laczen What if we add here run-time evaluation whether export_func is static_save_on?

@nvlsianpu, this is not needed. Even if in future the use of a export is extended again to provide a dump_to_screen, there is another possibility for the subsystem to protect against bad usage: provide a `h_export()` but do not use the provided export_function. This is exactly what @PavelVPV is doing, so I see no security problem with this.

Reply:
@Laczen so should this be reflected in the settings subsystem documentation?

Reply:
@PavelVPV, maybe, but the settings system has a lot of possibilities that are not explicitely described, this is a little difficult at times, but it is also extremely powerful (as shown by yourself). Are we still following the documentation: yes, is there a mentioning that it is obligatory to use the export_routine: no. I would document it in the mesh settings not in the settings, it is just clever usage of the settings module.